crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,327 @@
+      ctx->buf_len -= to_copy;
+      ctx->buf_off += to_copy;
+      bytes_output += to_copy;
+      remaining -= to_copy;
+      continue;  // continue to see if we've gotten |remaining| to 0
+    }
+    if (ctx->eof) {
+      break;
+    }
+    ctx->buf_len = 0;
+    ctx->buf_off = 0;
+    // |EVP_DecryptUpdate| may write up to cipher_block_size-1 more bytes than
+    // requested, so only read bytes we're sure we can decrypt in place.
+    int to_read = (int)sizeof(ctx->buf) - cipher_block_size + 1;
+    int bytes_read = BIO_read(next, &ctx->buf[ctx->buf_off], to_read);
+    if (bytes_read == 0 && BIO_eof(next)) {
+      ctx->eof = 1;
+      // EVP_CipherFinal_ex may write up to one block to our buffer. If that
+      // happens, continue the loop to process the decrypted block as normal.
+      ctx->ok = EVP_CipherFinal_ex(ctx->cipher, &ctx->buf[ctx->buf_off],
+                                   &ctx->buf_len);
+      if (ctx->buf_len > 0) {
+        continue;
+      }
+      return bytes_output;
+    } else if (bytes_read <= 0) {
+      BIO_copy_next_retry(b);
+      return bytes_output;
+    }
+    // Decrypt ciphertext in place, update |ctx->buf_len| with num bytes
+    // decrypted. Treat decrypt error as EOF and set error.
+    ctx->ok =
+        EVP_DecryptUpdate(ctx->cipher, &ctx->buf[ctx->buf_off], &ctx->buf_len,
+                          &ctx->buf[ctx->buf_off], bytes_read);
+    if (!ctx->ok) {
+      ctx->eof = 1;
+    }
+  }
+  return bytes_output;
+}
+
+static int enc_flush(BIO *b, BIO *next, BIO_ENC_CTX *ctx, int do_final) {
+  GUARD_PTR(b);
+  GUARD_PTR(next);
+  GUARD_PTR(ctx);
+  while (ctx->buf_len > 0) {
+    int outcome = BIO_write(next, &ctx->buf[ctx->buf_off], ctx->buf_len);
+    if (outcome <= 0) {
+      BIO_copy_next_retry(b);
+      return outcome;
+    }
+    ctx->buf_off += outcome;
+    ctx->buf_len -= outcome;
+  }
+  ctx->buf_off = 0;
+  ctx->buf_len = 0;
+  if (do_final && !ctx->flushed) {
+    ctx->flushed = 1;
+    ctx->ok = EVP_CipherFinal_ex(ctx->cipher, ctx->buf, &ctx->buf_len);
+    return ctx->ok;
+  }
+  return 1;
+}
+
+static int enc_write(BIO *b, const char *in, int inl) {
+  GUARD_PTR(b);
+  GUARD_PTR(in);
+  int ret = 0;
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+  BIO *next = BIO_next(b);
+  if (next == NULL) {
+    return 0;
+  }
+
+  if ((ret = enc_flush(b, next, ctx, /*do_final*/ 0)) <= 0) {
+    return ret;
+  }
+
+  assert(ctx->buf_off == ctx->buf_len && ctx->buf_len == 0);
+  int bytes_processed = 0;
+  int remaining = inl;
+  while (remaining > 0) {
+    if (ctx->buf_len == 0) {
+      ctx->buf_off = 0;
+      int to_encrypt =
+          remaining > (int)sizeof(ctx->buf) ? (int)sizeof(ctx->buf) : remaining;
+      if (!EVP_EncryptUpdate(ctx->cipher, &ctx->buf[0], &ctx->buf_len,
+                             (uint8_t *)in, to_encrypt)) {
+        ret = bytes_processed;
+        goto out;
+      };
+      bytes_processed += ctx->buf_len;
```

Comment:
The change to `bytes_processed` and `remaining` should be based on `to_encrypt`, i.e., the number of bytes passed as input, not the number of bytes generated as output (which could potentially be more than `remaining`).

Reply:
agreed. fixed and renamed `bytes_processed` for clarity.